### PR TITLE
Mc/hash vk

### PR DESCRIPTION
**Pull request by iAmMichaelConnor**
# Description

- Adding a method to hash the `verification_key_data` struct in a way which aligns with how the `verification_key` is hashed.
- (David will be adding tests :) )

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: cpp/src/barretenberg/proof_system/verification_key/verification_key.cpp [hunk 1]
```diff
@@ -1,10 +1,23 @@
...
+#include "barretenberg/polynomials/evaluation_domain.hpp"
 #include "verification_key.hpp"
 #include "../../plonk/proof_system/constants.hpp"
 
 namespace bonk {
 
+barretenberg::fr compress_native_evaluation_domain(barretenberg::evaluation_domain const& domain) {
+    barretenberg::fr out;
+
+    out = crypto::pedersen_commitment::compress_native({
```
Comment (kevaundray) [inline cpp/src/barretenberg/proof_system/verification_key/verification_key.cpp]: why is this needed?

Comment (iAmMichaelConnor) [reply cpp/src/barretenberg/proof_system/verification_key/verification_key.cpp]: I wasn't sure - I literally copied the exact way it's being done in cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp (see later file in this PR). This PR just seeks to align the two methods, so that the e2e tests pass.

File: cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp [hunk 2]
```diff
@@ -229,28 +229,30 @@ template <typename Curve> struct verification_key {
             return limbs;
         };
 
-        std::vector<barretenberg::fr> key_witnesses;
-        key_witnesses.push_back(compressed_domain);
-        key_witnesses.push_back(key->num_public_inputs);
+        std::vector<barretenberg::fr> preimage_data;
+        // TODO: put composer type in here.
```
Comment (iAmMichaelConnor) [inline cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp]: Remove comment

File: cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp [hunk 3]
```diff
@@ -229,28 +229,30 @@ template <typename Curve> struct verification_key {
...
+            preimage_data.push_back(y_limbs[2]);
+            preimage_data.push_back(y_limbs[3]);
         }
         barretenberg::fr compressed_key;
         if constexpr (Composer::type == ComposerType::PLOOKUP) {
-            compressed_key = crypto::pedersen_commitment::lookup::compress_native(key_witnesses);
+            compressed_key = crypto::pedersen_commitment::lookup::compress_native(preimage_data, hash_index);
         } else {
-            compressed_key = crypto::pedersen_commitment::compress_native(key_witnesses);
+            compressed_key = crypto::pedersen_commitment::compress_native(preimage_data, hash_index); // TODO: we need a hash index here!
```
Comment (iAmMichaelConnor) [inline cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp]: Remove comment

File: cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp [hunk 4]
```diff
@@ -188,7 +188,7 @@ template <typename Curve> struct verification_key {
     }
 
   public:
-    field_t<Composer> compress()
+    field_t<Composer> compress(size_t const hash_index = 0)
     {
         field_t<Composer> compressed_domain = domain.compress();
```
Comment (iAmMichaelConnor) [inline cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp:193]: Need to do `key_witnesses.push_back(Composer::type)` here - I forgot to add it (see the `compress_native` static method directly below, for reference)

Comment (iAmMichaelConnor) [reply cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp:193]: Also might be worth renaming `key_witnesses` to `preimage_data`

File: cpp/src/barretenberg/proof_system/verification_key/verification_key.cpp [hunk 5]
```diff
@@ -1,10 +1,23 @@
 #include "barretenberg/crypto/sha256/sha256.hpp"
 #include "barretenberg/crypto/pedersen_commitment/pedersen.hpp"
+#include "barretenberg/polynomials/evaluation_domain.hpp"
 #include "verification_key.hpp"
 #include "../../plonk/proof_system/constants.hpp"
 
 namespace bonk {
 
+barretenberg::fr compress_native_evaluation_domain(barretenberg::evaluation_domain const& domain) {
```
Comment (iAmMichaelConnor) [inline cpp/src/barretenberg/proof_system/verification_key/verification_key.cpp]: Forgot to add this comment:

```!
/**
 * @brief Hashes the evaluation domain to match the 'circuit' approach taken in stdlib/recursion/verification_key/verification_key.hpp.
 * @note: in that reference file, the circuit-equivalent of this function is a _method_ of the `evaluation_domain' struct. But we cannot do that with the native `barretenberg::evaluation_domain` type unfortunately, because it's defined in polynomials/evaluation_domain.hpp, and `polynomial` is a bberg library which does not depend on `crypto` in its CMakeLists.txt file. (We'd need `crypto` to be able to call native pedersen functions).
 * 
 * @param domain 
 * @return barretenberg::fr 
 */
```

File: cpp/src/barretenberg/proof_system/verification_key/verification_key.test.cpp [hunk 6]
```diff
@@ -1,43 +1,153 @@
...
 #include "verification_key.hpp"
 
+namespace {
+auto& engine = numeric::random::get_debug_engine();
+}
+
 using namespace barretenberg;
 using namespace bonk;
 
+/**
```
Comment (suyash67) [inline cpp/src/barretenberg/proof_system/verification_key/verification_key.test.cpp:13]: These tests are great! We should also test that `compress_native` in `cpp/src/barretenberg/proof_system/verification_key/verification_key.cpp` gives the same result as the `compress_native` in `cpp/src/barretenberg/stdlib/recursion/verification_key/verification_key.hpp`.

Comment (suyash67) [reply cpp/src/barretenberg/proof_system/verification_key/verification_key.test.cpp:13]: I've add a test. That should suffice for now.
